vispy/scene/canvas.py

```diff
@@ -241,73 +262,205 @@ def draw_visual(self, visual, event=None, viewport=None):
-                                                     TransformCache())
-        scene_event = SceneMouseEvent(canvas=self, event=event,
-                                      transform_cache=tr_cache)
-        scene_event.push_node(self.render_cs)
-        scene_event.push_node(self.framebuffer_cs)
-        scene_event.push_node(self.canvas_cs)
-        scene_event.push_node(self._scene)
-        prof('prepare mouse event')
+        prof = Profiler()  # noqa
+        deliver_types = ['mouse_press', 'mouse_wheel']
+        if self._send_hover_events:
+            deliver_types += ['mouse_move']
+
+        picked = self._mouse_handler
+        if picked is None:
+            if event.type in deliver_types:
+                picked = self.visual_at(event.pos)
+
+        # No visual to handle this event; bail out now
+        if picked is None:
+            return
         
-        self._scene._process_mouse_event(scene_event)
-        prof('process')
+        # Create an event to pass to the picked visual
+        scene_event = SceneMouseEvent(event=event, visual=picked)
         
+        # Deliver the event
+        if picked == self._mouse_handler:
+            # If we already have a mouse handler, then no other node may
+            # receive the event
+            if event.type == 'mouse_release':
+                self._mouse_handler = None
+            getattr(picked.events, event.type)(scene_event)
+        else:
+            # If we don't have a mouse handler, then pass the event through
+            # the chain of parents until a node accepts the event.
+            while picked is not None:
+                getattr(picked.events, event.type)(scene_event)
+                if scene_event.handled:
+                    if event.type == 'mouse_press':
+                        self._mouse_handler = picked
+                    break
+                if event.type in deliver_types:
+                    # events that are not handled get passed to parent
+                    picked = picked.parent
+                    scene_event.visual = picked
+                else:
+                    picked = None
+            
         # If something in the scene handled the scene_event, then we mark
         # the original event accordingly.
         event.handled = scene_event.handled
 
+    def visual_at(self, pos):
+        """Return the visual at a given position
+
+        Parameters
+        ----------
+        pos : tuple
+            The position in logical coordinates to query.
+
+        Returns
+        -------
+        visual : instance of Visual | None
+            The visual at the position, if it exists.
+        """
+        tr = self.transforms.get_transform('canvas', 'framebuffer')
+        fbpos = tr.map(pos)[:2]
+
+        try:
+            id_ = self._render_picking(region=(fbpos[0]-10, fbpos[1]-10,
+                                               20, 20))
+            vis = VisualNode._visual_ids.get(id_[10, 10], None)
```

Comment:
It's an artifact. I was playing with a few different methods of deciding which item was clicked on, but ultimately ended up using the separate `visuals_at` method instead. I expect these to evolve as we use them more.